### PR TITLE
Modifications for EVB v6

### PR DESCRIPTION
**Pull request by vmarandon**
Modifications to be able to read data from the EVB v6 while keeping the capability to read old files.

A storage class was added : NectarCAMDataStreamContainer which contain the information from TelescopeDataStream.
The scaling factors that are apply to the data imply a change of type from uint16 to float for the R0. It should be transparent though (and only done for v6 data)

A little bit of refactoring was done to avoid to have if/else at different place (so use the NectarCAMServiceContainer instead of the camera_config when possible)

I'm pinging also @sizun to have a look (or if he wants to test this branch) as it seems I can't assign him for the review.

### CODE REVIEW

File: src/ctapipe_io_nectarcam/__init__.py [hunk 1]
```diff
@@ -38,14 +38,16 @@
...
-    NectarCAMEventContainer
+    NectarCAMEventContainer, NectarCAMDataStreamContainer
 from .anyarray_dtypes import (
     CDTS_AFTER_37201_DTYPE,
     CDTS_BEFORE_37201_DTYPE,
     TIB_DTYPE,
 )
 from .version import __version__
 
+from IPython import embed
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:49]: Is this really needed? The only call to embed I see is in a block of code commented out?

File: src/ctapipe_io_nectarcam/__init__.py [hunk 2]
```diff
@@ -433,6 +435,15 @@ class NectarCAMEventSource(EventSource):
         help='If True, flat field and pedestal events are also calibrated.'
     ).tag(config=True)
 
+    def _correct_tel_id(self,tel_id,run):
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:438]: Add comments to explain what is done here? If I had to modify the code in the future I would not be able to understand it.

File: src/ctapipe_io_nectarcam/__init__.py [hunk 3]
```diff
@@ -622,7 +645,12 @@ def _generator(self):
             array_event.index.obs_id = self.obs_ids[0]
 
             # fill R0/R1 data
-            self.fill_r0r1_container(array_event, event)
+            self.fill_r0r1_container(array_event, event) ## svc or use self.nectarcam_service
+            
+            # bad_event = array_event.r0.tel[self.tel_id].waveform is None and array_event.r1.tel[self.tel_id].waveform is None
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:650]: What is this commented out code for? Clean up before merging?

File: src/ctapipe_io_nectarcam/__init__.py [hunk 4]
```diff
@@ -633,8 +661,9 @@ def _generator(self):
             # fill general monitoring data
             self.fill_mon_container_from_zfile(array_event, event)
 
+            
             # gain select and calibrate to pe
-            if self.r0_r1_calibrator.calibration_path is not None:
+            if self.r0_r1_calibrator.calibration_path is not None: # and not bad_event:
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:666]: Clean up here too?

File: src/ctapipe_io_nectarcam/__init__.py [hunk 5]
```diff
@@ -811,18 +908,21 @@ def fill_trigger_info(self, array_event):
...
 
-    def unpack_feb_data(self, event_container, event):
+    def unpack_feb_data(self, event_container, event, nectarcam_data):
         '''Unpack FEB counters and trigger pattern'''
 
         # Deduce data format version
         bytes_per_module = len(
-            event.nectarcam.counters) // self.camera_config.nectarcam.num_modules
+            nectarcam_data.counters) // self.nectarcam_service.num_modules
+        # VIM NOTES> There is in v6 the num_modules per event... Use this one ?
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:917]: I'm not sure, is this discussed in the documentation? Could this change if there are missing modules?

File: src/ctapipe_io_nectarcam/__init__.py [hunk 6]
```diff
@@ -872,15 +972,68 @@ def unpack_feb_data(self, event_container, event):
...
+                unpacked_charge = struct.unpack(rec_fmt, getattr(nectarcam_data,
                                                                  f'charges_gain{gain_id + 1}'))
                 event_container.native_charge[
-                    gain_id, self.camera_config.expected_pixels_id] = unpacked_charge
+                    gain_id, self.nectarcam_service.pixel_ids] = unpacked_charge
+
+        if not self.pre_v6_data:
+            event_container.first_cell_id = np.full( (N_PIXELS,), -1, dtype=event.first_cell_id.dtype)
+            event_container.first_cell_id[ self.nectarcam_service.pixel_ids ] = event.first_cell_id
+        #embed()
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:989]: Is this code meant to be added to the PR before merging?

File: src/ctapipe_io_nectarcam/__init__.py [hunk 7]
```diff
@@ -891,29 +1044,52 @@ def fill_r0r1_camera_container(self, zfits_event):
...
+        else:
+            dtype = np.float32 
+            fill = np.finfo(dtype).max
+        # VIM : We received uint16 data from the EVB, 
+        # but because of the scaling in v6 we gets back to a float, so let's use directly floats
+
         # we assume that either all pixels are gain selected or none
         # only broken pixels are allowed to be missing completely
+
+        # if all_broken:
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:1080]: Is this code meant to be added to the PR before merging or should it be cleaned up?

File: src/ctapipe_io_nectarcam/calibration.py [hunk 8]
```diff
@@ -159,13 +161,13 @@ def _read_calibration_file(path):
...
                 if key.startswith('tel_')
             ]
-
         for tel_id in tel_ids:
             with HDF5TableReader(path) as h5_table:
-
                 mon.tel[tel_id] = MonitoringCameraContainer(
                     calibration=next(h5_table.read(f'/tel_{tel_id}/calibration', WaveformCalibrationContainer)),
                     flatfield=next(h5_table.read(f'/tel_{tel_id}/flatfield', FlatFieldContainer)),
+                    #calibration=next(h5_table.read(f'/tel_{tel_id_hack}/calibration', WaveformCalibrationContainer)),
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/calibration.py:169]: Clean up?

File: src/ctapipe_io_nectarcam/containers.py [hunk 9]
```diff
@@ -10,6 +10,15 @@
     'NectarCAMCameraContainer',
 ]
 
+class NectarCAMDataStreamContainer(Container):
+    """
+    Container for the Fields that are from the DataStream information (new from v6)
+    """
+    telescope_id = Field(-1,"telescope id")
+    sb_id = Field(-1,"No idea what this is")
```
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/containers.py:18]: ask Dirk?

File: src/ctapipe_io_nectarcam/__init__.py (hunk 2 above)
Comment (tibaldo) [inline src/ctapipe_io_nectarcam/__init__.py:438]: Make static method and move after init?

**Review by tibaldo — COMMENTED** [review]
Thank you for taking care of this @vmarandon !
General comments.
- There are several blocks of code commented out. I'm not sure if they are there to bbe completed in the future or if they should be cleaned up before merging.
- I think it would be extremely useful to add a small test file and expand the unit tests to include checks on ingesting v6 data. This will help to avoid breaking the reader in future PR.

File: src/ctapipe_io_nectarcam/containers.py (hunk 9 above)
Comment (kosack) [reply src/ctapipe_io_nectarcam/containers.py:18]: `sb_id` is the scheduling block ID that you get from ACADA.  This should not be stored here in any case, but in your EventSource's `scheduling_blocks` property which should be a `dict[sb_id, ctapipe.containers.SchedulingBlockContainer]`, similar to the `observation_blocks` property for `obs_ids`.   Most of what is in DataStream is not really applicable to a single telescope operating without ACADA, so I don't think you need this container, unless you are using it to temporarily store some information before setting up the EventSource. 

If you don't have an sb_id, you can just set it to some dummy value or to the same value as the obs_id, as we do in the SimTelEventSource currently.

File: src/ctapipe_io_nectarcam/__init__.py (hunk 1 above)
Comment (vmarandon) [reply src/ctapipe_io_nectarcam/__init__.py:49]: No, it's a remaining of some debugging process. I've removed it.
I was too hurry in the commit process in seems 😅

File: src/ctapipe_io_nectarcam/__init__.py (hunk 2 above)
Comment (vmarandon) [reply src/ctapipe_io_nectarcam/__init__.py:438]: Yes. I've now added some comments in the code about that.

File: src/ctapipe_io_nectarcam/containers.py (hunk 9 above)
Comment (vmarandon) [reply src/ctapipe_io_nectarcam/containers.py:18]: Thanks for the clarification on those variables. 
The container class is there to store temporary the information. I decided to go for a container to do as the other class which allow to have it along the data for debugging purpose.

File: src/ctapipe_io_nectarcam/__init__.py (hunk 2 above)
Comment (vmarandon) [reply src/ctapipe_io_nectarcam/__init__.py:438]: As the only purpose is only to be applied to this class, what would be the purpose of making it a static method ? (I'm not an expert of python, so I don't know all the good practice and techniques)

**Review by jlenain — APPROVED** [review]
Huge thanks for this PR, @vmarandon !
A small run taken with the _EventBuilder_ version 6 has been uploaded to the test server. Locally, all unit tests successfully pass, ensuring compatibility with _EventBuilder_ version 6 and backward-compatibility with _EventBuilder_ version 4.
In my opinion, the draft status of this PR can be lifted, the CI tests can be triggered again on GitHub, and if successful, we'll just need to accommodate for small conflicts following the merge of PR #47 , and this PR would be ready for merge.

**Comment by vmarandon** [comment]
I've merged the branch with the master according to the change in LightNectarCAMEventSource discussed on slack.
The conflict seems to be solved but would be nice if you could have a look to see if for you everything looks ok. (I've just been through it and seems ok to me)
If so, we should then merge the pull request.

**Comment by jlenain** [comment]
Thanks a lot, @vmarandon ! I very much like your re-implementation of `LightNectarCAMEventSource`, which is super elegant.

I also went through your last changes, and tested everything locally. I just pushed a small commit increasing slightly the coverage on tests for the light version of the event source.

If the CI tests pass, all would seem good to me too for merging this PR.

**Review by jlenain — APPROVED** [review]
Many thanks again, @vmarandon !

**Pull request merged**